Installation issue: CMake using Intel compilers

Using `spack@develop` (96fa6f0c1be4ab55ec6ba7cd5af059e1ed95351f, from 17-July-2020) at NERSC Cori on `haswell`:

Building `CMake@3.17.3` fails with:
1. `%intel@19.0.3.199` -- fails in boostrap phase (see below)
2. `%intel@19.1.1.217` -- fails in bootstrap phase (same error as 1.)
3. `%intel@18.0.3.222` -- fails during build (see below)


**%intel@19.0.3.199, %intel@19.1.1.217 -- bootstrap fails**
```
$> spack install cmake@3.17.3%intel@19.0.3.199
...
==> 21896: Installing cmake
...
==> No patches needed for cmake
==> 21896: cmake: Building cmake [Package]
==> 21896: cmake: Executing phase: 'bootstrap'
...
==> Error: ProcessError: Command exited with status 7:
    './bootstrap' '--prefix=/global/u1/l/eugenewalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.0.3.199/cmake-3.17.3-ajy6uihnh2s34y6uwnkagznxvygjb2rm' '--parallel=16' '--no-system-libs' '--no-qt-gui' '--' '-DCMAKE_BUILD_TYPE=Release' '-DCMake_TEST_INSTALL=OFF' '-DCMAKE_USE_OPENSSL=True'
1 error found in build log:
     1    ==> 21896: cmake: Executing phase: 'bootstrap'
     2    ==> [2020-07-20-12:04:55.896236, 21967] './bootstrap' '--prefix=/global/u1/l/eugenewalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placehold
          er/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeh
          older/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_pla
          ceholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.
          0.3.199/cmake-3.17.3-ajy6uihnh2s34y6uwnkagznxvygjb2rm' '--parallel=16' '--no-system-libs' '--no-qt-gui' '--' '-DCMAKE_BUILD_TYPE=Release' '-DCMake_TEST_INSTAL
          L=OFF' '-DCMAKE_USE_OPENSSL=True'
     3    ---------------------------------------------
     4    CMake 3.17.3, Copyright 2000-2020 Kitware, Inc. and Contributors
     5    C compiler on this system is: /global/u1/l/eugenewalker/spack/lib/spack/env/intel/icc
     6    ---------------------------------------------
  >> 7    Error when bootstrapping CMake:
     8    Cannot find a C++ compiler that supports both C++11 and the specified C++ flags.
     9    Please specify one using environment variable CXX.
     10   The C++ flags are "".
     11   They can be changed using the environment variable CXXFLAGS.
     12   See cmake_bootstrap.log for compilers attempted.
     13   ---------------------------------------------
```

**%intel@18.0.3.222 -- build fails**
```
$> spack install cmake@3.17.3%intel@18.0.3.222
==> 49654: Installing cmake
...
==> No patches needed for cmake
==> 49654: cmake: Building cmake [Package]
==> 49654: cmake: Executing phase: 'bootstrap'
==> 49654: cmake: Executing phase: 'build'
...
21 errors found in build log:
     1989    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/cmArchiveWrite.cxx.o
     1990    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/cmBase32.cxx.o
     1991    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/cmArgumentParser.cxx.o
     1992    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/cmBinUtilsLinuxELFLinker.cxx.o
     1993    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArgumentParser.h(9),
     1994                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArgumentParser.cxx(3):
  >> 1995    /usr/include/c++/7/functional(139): error: class template "std::_Mem_fn_traits" has already been defined
     1996      _GLIBCXX_MEM_FN_TRAITS(noexcept, true_type, true_type)
     1997      ^
     1998
     1999    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArgumentParser.h(9),
     2000                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArgumentParser.cxx(3):
  >> 2001    /usr/include/c++/7/functional(140): error: class template "std::_Mem_fn_traits" has already been defined
     2002      _GLIBCXX_MEM_FN_TRAITS(& noexcept, true_type, false_type)
     2003      ^
     2004
     2005    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArgumentParser.h(9),
     2006                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArgumentParser.cxx(3):
  >> 2007    /usr/include/c++/7/functional(141): error: class template "std::_Mem_fn_traits" has already been defined
     2008      _GLIBCXX_MEM_FN_TRAITS(&& noexcept, false_type, true_type)
     2009      ^
     2010
     2011    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmRange.h(9),
     2012                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmStringAlgorithms.h(18),
     2013                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArchiveWrite.cxx(18):
  >> 2014    /usr/include/c++/7/functional(139): error: class template "std::_Mem_fn_traits" has already been defined
     2015      _GLIBCXX_MEM_FN_TRAITS(noexcept, true_type, true_type)
     2016      ^
     2017
     2018    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmRange.h(9),
     2019                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmStringAlgorithms.h(18),
     2020                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArchiveWrite.cxx(18):
  >> 2021    /usr/include/c++/7/functional(140): error: class template "std::_Mem_fn_traits" has already been defined
     2022      _GLIBCXX_MEM_FN_TRAITS(& noexcept, true_type, false_type)
     2023      ^
     2024
     2025    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmRange.h(9),
     2026                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmStringAlgorithms.h(18),
     2027                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArchiveWrite.cxx(18):
  >> 2028    /usr/include/c++/7/functional(141): error: class template "std::_Mem_fn_traits" has already been defined
     2029      _GLIBCXX_MEM_FN_TRAITS(&& noexcept, false_type, true_type)
     2030      ^
     2031
     2032    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/cmBinUtilsLinuxELFObjdumpGetRuntimeDependenciesTool.cxx.o
     2033    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/cmBinUtilsMacOSMachOLinker.cxx.o
     2034    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/cmBinUtilsMacOSMachOGetRuntimeDependenciesTool.cxx.o
     2035    compilation aborted for /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArgumentParser.cxx (code 2)
  >> 2036    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/build.make:239: Source/CMakeFiles/CMakeLib.dir/cmArgumentParser.cxx.o] Error 2
     2037    make[2]: *** Waiting for unfinished jobs....
     2038    compilation aborted for /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmArchiveWrite.cxx (code 2)
     2039    [ 59%] Building CXX object Source/CMakeFiles/CMakeLib.dir/cmBinUtilsMacOSMachOOToolGetRuntimeDependenciesTool.cxx.o
  >> 2040    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/build.make:226: Source/CMakeFiles/CMakeLib.dir/cmArchiveWrite.cxx.o] Error 2
     2041    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmMakefile.h(10),
     2042                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsLinuxELFLinker.cxx(14)
             :
  >> 2043    /usr/include/c++/7/functional(139): error: class template "std::_Mem_fn_traits" has already been defined
     2044      _GLIBCXX_MEM_FN_TRAITS(noexcept, true_type, true_type)
     2045      ^
     2046
     2047    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmMakefile.h(10),
     2048                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsLinuxELFLinker.cxx(14)
             :
  >> 2049    /usr/include/c++/7/functional(140): error: class template "std::_Mem_fn_traits" has already been defined
     2050      _GLIBCXX_MEM_FN_TRAITS(& noexcept, true_type, false_type)
     2051      ^
     2052
     2053    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmMakefile.h(10),
     2054                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsLinuxELFLinker.cxx(14)
             :
  >> 2055    /usr/include/c++/7/functional(141): error: class template "std::_Mem_fn_traits" has already been defined
     2056      _GLIBCXX_MEM_FN_TRAITS(&& noexcept, false_type, true_type)
     2057      ^
     2058
     2059    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmSystemTools.h(9),
     2060                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsLinuxELFObjdumpGetRunt
             imeDependenciesTool.cxx(11):
  >> 2061    /usr/include/c++/7/functional(139): error: class template "std::_Mem_fn_traits" has already been defined
     2062      _GLIBCXX_MEM_FN_TRAITS(noexcept, true_type, true_type)
     2063      ^
     2064
     2065    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmSystemTools.h(9),
     2066                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsLinuxELFObjdumpGetRunt
             imeDependenciesTool.cxx(11):
  >> 2067    /usr/include/c++/7/functional(140): error: class template "std::_Mem_fn_traits" has already been defined
     2068      _GLIBCXX_MEM_FN_TRAITS(& noexcept, true_type, false_type)
     2069      ^
     2070
     2071    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmSystemTools.h(9),
     2072                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsLinuxELFObjdumpGetRunt
             imeDependenciesTool.cxx(11):
  >> 2073    /usr/include/c++/7/functional(141): error: class template "std::_Mem_fn_traits" has already been defined
     2074      _GLIBCXX_MEM_FN_TRAITS(&& noexcept, false_type, true_type)
     2075      ^
     2076
     2077    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmRange.h(9),
     2078                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmStringAlgorithms.h(18),
     2079                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsMacOSMachOLinker.cxx(1
             4):
  >> 2080    /usr/include/c++/7/functional(139): error: class template "std::_Mem_fn_traits" has already been defined
     2081      _GLIBCXX_MEM_FN_TRAITS(noexcept, true_type, true_type)
     2082      ^
     2083
     2084    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmRange.h(9),
     2085                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmStringAlgorithms.h(18),
     2086                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsMacOSMachOLinker.cxx(1
             4):
  >> 2087    /usr/include/c++/7/functional(140): error: class template "std::_Mem_fn_traits" has already been defined
     2088      _GLIBCXX_MEM_FN_TRAITS(& noexcept, true_type, false_type)
     2089      ^
     2090
     2091    In file included from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmRange.h(9),
     2092                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmStringAlgorithms.h(18),
     2093                     from /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsMacOSMachOLinker.cxx(1
             4):
  >> 2094    /usr/include/c++/7/functional(141): error: class template "std::_Mem_fn_traits" has already been defined
     2095      _GLIBCXX_MEM_FN_TRAITS(&& noexcept, false_type, true_type)
     2096      ^
     2097
     2098    compilation aborted for /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsMacOSMachOLinker.cxx
              (code 2)
  >> 2099    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/build.make:330: Source/CMakeFiles/CMakeLib.dir/cmBinUtilsMacOSMachOLinker.cxx.o] Error 2
     2100    compilation aborted for /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsLinuxELFObjdumpGetRu
             ntimeDependenciesTool.cxx (code 2)
  >> 2101    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/build.make:304: Source/CMakeFiles/CMakeLib.dir/cmBinUtilsLinuxELFObjdumpGetRuntimeDependenciesTool.cxx.o] Erro
             r 2
     2102    compilation aborted for /tmp/eugenewalker/spack-stage/spack-stage-cmake-3.17.3-p5mydnmatmfo2arl3yzxkcjqtpszz7gr/spack-src/Source/cmBinUtilsLinuxELFLinker.cxx (
             code 2)
  >> 2103    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/build.make:291: Source/CMakeFiles/CMakeLib.dir/cmBinUtilsLinuxELFLinker.cxx.o] Error 2
  >> 2104    make[1]: *** [CMakeFiles/Makefile2:2412: Source/CMakeFiles/CMakeLib.dir/all] Error 2
     2105    make: *** [Makefile:183: all] Error 2
...
```

Full build logs:
* `%intel@19.0.3.199` - [cmake-intel.txt](https://github.com/spack/spack/files/4949615/cmake-intel.txt)
* `%intel@18.0.3.222` - [cmake-intel-18.txt](https://github.com/spack/spack/files/4949933/cmake-intel-18.txt)


@becker33 @shahzebsiddiqui

--- a/var/spack/repos/builtin/packages/cmake/intel-cxx-bootstrap.patch
+++ b/var/spack/repos/builtin/packages/cmake/intel-cxx-bootstrap.patch
@@ -1,0 +1,11 @@
+--- a/bootstrap	2020-07-24 11:02:56.488414260 +0200
++++ b/bootstrap	2020-07-24 11:03:20.763775094 +0200
+@@ -1179,7 +1179,7 @@
+ }
+ #endif
+ 
+-#if (__cplusplus >= 201703L || defined(__INTEL_COMPILER) && defined(__cpp_if_constexpr))
++#if (__cplusplus >= 201703L || defined(__INTEL_COMPILER) && defined(__cpp_deduction_guides))
+ #include <optional>
+ template <typename T,
+           typename std::invoke_result<decltype(&T::get), T>::type = nullptr>

--- a/var/spack/repos/builtin/packages/cmake/package.py
+++ b/var/spack/repos/builtin/packages/cmake/package.py
@@ -136,6 +136,10 @@ class Cmake(Package):
     # https://gitlab.kitware.com/cmake/cmake/issues/16226
     patch('intel-c-gnu11.patch', when='@3.6.0:3.6.1')
 
+    # Cannot build with Intel again, should be fixed in 3.17.4 and 3.18.1
+    # https://gitlab.kitware.com/cmake/cmake/-/issues/21013
+    patch('intel-cxx-bootstrap.patch', when='@3.17.0:3.17.3,3.18.0')
+
     # https://gitlab.kitware.com/cmake/cmake/issues/18232
     patch('nag-response-files.patch', when='@3.7:3.12')
 

